mlir/include/mlir/IR/TypeRange.h

```diff
@@ -47,6 +47,7 @@ class TypeRange : public llvm::detail::indexed_accessor_range_base<
   template <typename Arg, typename = std::enable_if_t<std::is_constructible<
                               ArrayRef<Type>, Arg>::value>>
   TypeRange(Arg &&arg) : TypeRange(ArrayRef<Type>(std::forward<Arg>(arg))) {}
+  TypeRange(std::nullopt_t) : TypeRange(ArrayRef<Type>()) {}
```

Comment:
For `TypeRange` and `ValueRange` I opted to allow clients to continue to pass in `std::nullopt` to get an empty range, to avoid chasing too far down the rabbit hole.